File: CS3315/Assignments/Week4Assignment/scripts/topWords.pl
#!/usr/bin/env perl

# open the file
open($file, '<', 'HuckFinn.txt');

# read contents of file into memory
@whole_file = <$file>;

# close file since its contents is now in memory
close($file);

# this is a hashmap that will hold key, value pairs
# where the key is the word and the value is the
# number of times that it appears
%data = ();

foreach $line (@whole_file) {
    # convert each line to lowercase
    $line = lc $line;
    
    # find all substrings that match the regular expression
    foreach $word ($line =~ /[a-z']+/g) {
        # update the number of times that it has appeared so far
        if (exists($data{$word})) {
            $data{$word} = $data{$word} + 1;
        } else {
            $data{$word} = 1;
        }
    }
}

# sort the key, value pairs based on their values
@top_ten = (reverse sort {$data{$a} <=> $data{$b}} keys %data)[0..9];

# display the top ten
foreach $key (@top_ten) {
    print "   $data{$key} $key\n";
}

File: CS3315/Assignments/Week4Assignment/scripts/topWords.lua
#!/usr/bin/env lua

local stopwords = {}
for line in io.lines('stopwords.txt') do
    stopwords[line] = line
end

-- this is a hashmap that will hold key, value pairs
-- where the word is the key and the number of times
-- that it appears is its value
local freq = {}

local site = io.read('*a')

-- convert each line of the text to lowercase
local lowerline = string.lower(site)

-- find all substrings that match the regular expression
for word in string.gmatch(lowerline, "[%a']+") do
    -- update the number of times that the substring
    -- has appeared so far
    if (not freq[word] and not stopwords[word]) then
	freq[word] = 1
    elseif freq[word] then
	freq[word] = freq[word] + 1
    end
end

local array = {}

-- fill array with the key, value pairs. I.E. array[0] => 'and, 6440'
-- this is necessary so that we can sort them correctly
for word, count in pairs(freq) do
	table.insert(array, {word, count})
end

-- sort the key, value pairs based on their value
table.sort(array, function (a, b) return a[2] > b[2] end)
 
-- display the top ten
for i = 1, 10 do
	io.write(string.format('%7d %s\n', array[i][2] , array[i][1]))
end
